### PR TITLE
cmd: Implement image import and export commands

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Implement commands to export / import gadget images to / from a file.

### Why 

Being able to export and export images from files is useful for the following cases:
- It's possible to distribute images by just using files. Airgapped environments. Cases where reaching the internet is not an option
- Integrate gadget images directly into the application binary. #2021. (I already have a POC of this working, but it's still missing a lot of things)

### Testing

```bash 
# let's start with an empty image store
$ sudo rm -rf /var/lib/ig/oci-store
$ sudo -E ig image list
INFO[0000] Experimental features enabled
REPOSITORY             TAG                    DIGEST       CREATED


# pull one image (build will work as well)
$ sudo -E ig image pull trace_open
INFO[0000] Experimental features enabled
Pulling trace_open...
Successfully pulled ghcr.io/inspektor-gadget/gadget/trace_open:latest@sha256:1ef0b57afaafe4e7d1befffc902cdda56a3d51936b9301710047b722f8018145

# export image
$ sudo -E ig image export trace_open /tmp/images.tar 
INFO[0000] Experimental features enabled
Successfully exported images to /tmp/images.tar

# clean the image store (or copy tar file to another machine...)
$ sudo rm -rf /var/lib/ig/oci-store
$ sudo -E ig image list
INFO[0000] Experimental features enabled
REPOSITORY             TAG                    DIGEST       CREATED

# import the image
$ sudo -E ig image import /tmp/images.tar
INFO[0000] Experimental features enabled
Successfully imported images:
  ghcr.io/inspektor-gadget/gadget/trace_open:latest

#  image is there and can be run
$ sudo -E ig image list
INFO[0000] Experimental features enabled
REPOSITORY             TAG                    DIGEST       CREATED
trace_open             latest                 1ef0b57afaaf 56 minutes ago

$ sudo -E ig run trace_open --pull never
INFO[0000] Experimental features enabled
RUNTIME.C… PID     UID      GID      MNT… … FD     … MODE  COMM  FNAME      TIME

```

### CODE REVIEW

**Review by mqasimsarfraz — COMMENTED** [review]
I tested it and it seems to be working fine. Just one comment regarding `import` command. 

Also one question regarding integration tests. I think we should drop the test for [build command](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/integration/components/image/image_test.go#L75) and use import/export to simplify (get rid of dependency of ebpf-builder image) them and also to test this new functionality? (can be handled in a seperate PR as well)

File: pkg/oci/oci.go [hunk 1]
```diff
@@ -234,6 +235,117 @@ func TagGadgetImage(ctx context.Context, srcImage, dstImage string) (*GadgetImag
...
+	if err := tarFolderToFile(tmpDir, dstFile); err != nil {
+		return fmt.Errorf("creating tar for gadget image: %w", err)
+	}
+
+	return nil
+}
+
+// ImportGadgetImages imports all the tagged gadget images from the src file.
+func ImportGadgetImages(ctx context.Context, srcFile string) ([]string, error) {
+	src, err := oci.NewFromTar(ctx, srcFile)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/oci/oci.go:276]: Good question. The same happens for the pull command

```bash
$ sudo -E ig image pull docker.io/busybox
INFO[0000] Experimental features enabled
Pulling docker.io/busybox...
Successfully pulled docker.io/library/busybox:latest@sha256:c3839dd800b9eb7603340509769c43e146a74c63dca3045a8e7dc8ee07e53966

$ sudo -E ig image list
INFO[0000] Experimental features enabled
REPOSITORY                                TAG                                      DIGEST       CREATED
docker.io/library/busybox                 latest                                   c3839dd800b9
```

The OCI logic doesn't check if the image being pulled / imported is actually a gadget image. (It'll require to check the layers of the image and so on). I think it's fine for now.

**Comment by mauriciovasquezbernal** [comment]
> Also one question regarding integration tests. I think we should drop the test for [build command](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/integration/components/image/image_test.go#L75) and use import/export to simplify (get rid of dependency of ebpf-builder image) them and also to test this new functionality? (can be handled in a seperate PR as well)

I'm wondering where can we import the image from, I mean, we can remove the dependency from the build command, but then we'll need to have an image to import somewhere.

File: pkg/oci/oci.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply pkg/oci/oci.go:276]: I opened this: https://github.com/inspektor-gadget/inspektor-gadget/issues/2788

**Comment by mqasimsarfraz** [comment]
> > Also one question regarding integration tests. I think we should drop the test for [build command](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/integration/components/image/image_test.go#L75) and use import/export to simplify (get rid of dependency of ebpf-builder image) them and also to test this new functionality? (can be handled in a seperate PR as well)
> 
> I'm wondering where can we import the image from, I mean, we can remove the dependency from the build command, but then we'll need to have an image to import somewhere.

That is good question, I see following options:

1. Export an image for once and put in `testdata` since we don't care about content of image
2. Keep the build command tests and use it to validate the import/export command.

Perhaps 2 is better?

**Review by mqasimsarfraz — APPROVED** [review]
LGTM! I am fine with adding tests in follow up PR! 

I like it that it works fine with multiple images:

```
→ go run -exec "sudo -E" ./cmd/ig image export trace_exec trace_open gadgets.tar
INFO[0000] Experimental features enabled                
Successfully exported images to gadgets.tar
→ go run -exec "sudo -E" ./cmd/ig image import gadgets.tar
INFO[0000] Experimental features enabled                
Successfully imported images:
  ghcr.io/inspektor-gadget/gadget/trace_exec:latest
  ghcr.io/inspektor-gadget/gadget/trace_open:latest
→ go run -exec "sudo -E" ./cmd/ig image list 
INFO[0000] Experimental features enabled                
REPOSITORY                                                              TAG                                                                     DIGEST       CREATED                                                                
trace_exec                                                              latest                                                                  2d559bee7bfe 37 minutes ago                                                         
trace_open                                                              latest                                                                  25ef26e76280 19 hours ago   
```

**Comment by flyth** [comment]
Great addition!

It might make sense to (in a follow-up PR) move the importing functionality to the runtime and thus also add gRPC calls for it. That way you could upload gadgets to ig-k8s as well. That'd still not be a perfect solution, as newly added nodes wouldn't receive the gadget that way, but it might be helpful in airgapped clusters that only need to run gadgets in an ad-hoc way on specific nodes.

**Pull request merged**